File: Backend/src/main/java/com/sencarmarket/module/avis/controller/AvisController.java
package com.sencarmarket.module.avis.controller;

import com.sencarmarket.module.avis.dto.AvisResponse;
import com.sencarmarket.module.avis.dto.CreateAvisRequest;
import com.sencarmarket.module.avis.service.AvisService;
import com.sencarmarket.module.commun.dto.PaginatedResponse;
import com.sencarmarket.module.commun.security.AccessControlService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.UUID;

/**
 * Controleur pour la gestion des avis
 */
@RestController
@RequestMapping("/api/avis")
@RequiredArgsConstructor
@PreAuthorize("isAuthenticated()")
public class AvisController {

    private final AvisService avisService;
    private final AccessControlService accessControlService;

    /**
     * Creer un nouvel avis - Utilisateurs authentifies uniquement
     */
    @PostMapping
    public ResponseEntity<AvisResponse> createAvis(
            @Valid @RequestBody CreateAvisRequest request,
            Authentication authentication) {
        UUID utilisateurId = accessControlService.getCurrentUserId(authentication);
        AvisResponse response = avisService.createAvis(request, utilisateurId);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    /**
     * Obtenir un avis par son ID
     */
    @GetMapping("/{avisId}")
    public ResponseEntity<AvisResponse> getAvisById(@PathVariable UUID avisId) {
        AvisResponse response = avisService.getAvisById(avisId);
        return ResponseEntity.ok(response);
    }

    /**
     * Obtenir les avis sur un utilisateur
     */
    @GetMapping("/utilisateur/{utilisateurId}")
    public ResponseEntity<PaginatedResponse<AvisResponse>> getAvisByUtilisateur(
            @PathVariable UUID utilisateurId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        PaginatedResponse<AvisResponse> response = avisService.getAvisByUtilisateur(utilisateurId, page, size);
        return ResponseEntity.ok(response);
    }

    /**
     * Obtenir les avis sur un véhicule
     */
    @GetMapping("/vehicule/{vehiculeId}")
    public ResponseEntity<PaginatedResponse<AvisResponse>> getAvisByVehicule(
            @PathVariable UUID vehiculeId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        PaginatedResponse<AvisResponse> response = avisService.getAvisByVehicule(vehiculeId, page, size);
        return ResponseEntity.ok(response);
    }

    /**
     * Obtenir les avis sur un garage
     */
    @GetMapping("/garage/{garageId}")
    public ResponseEntity<PaginatedResponse<AvisResponse>> getAvisByGarage(
            @PathVariable UUID garageId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size) {
        PaginatedResponse<AvisResponse> response = avisService.getAvisByGarage(garageId, page, size);
        return ResponseEntity.ok(response);
    }

    /**
     * Obtenir la note moyenne d'un utilisateur
     */
    @GetMapping("/utilisateur/{utilisateurId}/moyenne")
    public ResponseEntity<Double> getNoteMoyenneUtilisateur(@PathVariable UUID utilisateurId) {
        Double moyenne = avisService.getNoteMoyenneUtilisateur(utilisateurId);
        return ResponseEntity.ok(moyenne);
    }

    /**
     * Obtenir la note moyenne d'un véhicule
     */
    @GetMapping("/vehicule/{vehiculeId}/moyenne")
    public ResponseEntity<Double> getNoteMoyenneVehicule(@PathVariable UUID vehiculeId) {
        Double moyenne = avisService.getNoteMoyenneVehicule(vehiculeId);
        return ResponseEntity.ok(moyenne);
    }

    /**
     * Obtenir la note moyenne d'un garage
     */
    @GetMapping("/garage/{garageId}/moyenne")
    public ResponseEntity<Double> getNoteMoyenneGarage(@PathVariable UUID garageId) {
        Double moyenne = avisService.getNoteMoyenneGarage(garageId);
        return ResponseEntity.ok(moyenne);
    }

    /**
     * Signaler un avis
     */
    @PostMapping("/{avisId}/signaler")
    public ResponseEntity<Void> signalerAvis(
            @PathVariable UUID avisId,
            Authentication authentication) {
        UUID utilisateurId = accessControlService.getCurrentUserId(authentication);
        avisService.signalerAvis(avisId, utilisateurId);
        return ResponseEntity.ok().build();
    }

    /**
     * Supprimer un avis
     */
    @DeleteMapping("/{avisId}")
    public ResponseEntity<Void> deleteAvis(
            @PathVariable UUID avisId,
            Authentication authentication) {
        UUID utilisateurId = accessControlService.getCurrentUserId(authentication);
        avisService.deleteAvis(avisId, utilisateurId);
        return ResponseEntity.noContent().build();
    }

    /**
     * Vérifier si une transaction est valide pour laisser un avis
     */
    @GetMapping("/transaction/{transactionId}/validation")
    public ResponseEntity<Boolean> isTransactionValide(
            @PathVariable UUID transactionId,
            @RequestParam String typeAvis) {
        boolean valide = avisService.isTransactionValide(transactionId, typeAvis);
        return ResponseEntity.ok(valide);
    }

}

File: Backend/src/modules/avis/avis.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseIntPipe,
  ParseUUIDPipe,
  Post,
  Query,
  UseGuards,
} from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiExtraModels,
  ApiOperation,
  ApiParam,
  ApiQuery,
  ApiResponse,
  ApiTags,
  getSchemaPath,
} from '@nestjs/swagger';

import { CurrentUser } from '../../common/decorators/current-user.decorator';
import { PaginatedResponseDto } from '../../common/dto/paginated-response.dto';
import { JwtAuthGuard } from '../../common/guards/jwt-auth.guard';
import type { AuthenticatedUser } from '../../common/types/authenticated-user.type';
import { ApiErrorResponseDto } from '../auth/dto/api-error-response.dto';

import { AvisService } from './avis.service';
import { AvisResponseDto } from './dto/avis-response.dto';
import { CreateAvisRequestDto } from './dto/create-avis-request.dto';

@ApiTags('Avis et Notes')
@UseGuards(JwtAuthGuard)
@ApiBearerAuth()
@ApiExtraModels(PaginatedResponseDto, AvisResponseDto)
@Controller('avis')
export class AvisController {
  constructor(private readonly service: AvisService) {}

  @Post()
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: 'Créer un avis' })
  @ApiResponse({ status: 201, type: AvisResponseDto, description: 'Avis créé avec succès' })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Données invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Entité non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  createAvis(
    @Body() request: CreateAvisRequestDto,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<AvisResponseDto> {
    return this.service.createAvis(request, user);
  }

  @Get()
  @ApiOperation({ summary: 'Obtenir tous les avis' })
  @ApiQuery({ name: 'page', required: false, type: Number, example: 0 })
  @ApiQuery({ name: 'size', required: false, type: Number, example: 10 })
  @ApiResponse({
    status: 200,
    description: 'Avis récupérés avec succès',
    schema: {
      allOf: [
        { $ref: getSchemaPath(PaginatedResponseDto) },
        { properties: { content: { type: 'array', items: { $ref: getSchemaPath(AvisResponseDto) } } } },
      ],
    },
  })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getAllAvis(
    @Query('page', new ParseIntPipe({ optional: true })) page = 0,
    @Query('size', new ParseIntPipe({ optional: true })) size = 10,
  ): Promise<PaginatedResponseDto<AvisResponseDto>> {
    return this.service.getAllAvis(page, size);
  }

  @Get(':avisId')
  @ApiOperation({ summary: 'Obtenir un avis par ID' })
  @ApiParam({ name: 'avisId', type: String, format: 'uuid' })
  @ApiResponse({ status: 200, type: AvisResponseDto, description: 'Avis récupéré avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Avis non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getAvisById(@Param('avisId', new ParseUUIDPipe()) avisId: string): Promise<AvisResponseDto> {
    return this.service.getAvisById(avisId);
  }

  @Get('utilisateur/:utilisateurId')
  @ApiOperation({ summary: 'Avis sur un utilisateur' })
  @ApiParam({ name: 'utilisateurId', type: String, format: 'uuid' })
  @ApiQuery({ name: 'page', required: false, type: Number, example: 0 })
  @ApiQuery({ name: 'size', required: false, type: Number, example: 10 })
  @ApiResponse({
    status: 200,
    description: 'Avis récupérés avec succès',
    schema: {
      allOf: [
        { $ref: getSchemaPath(PaginatedResponseDto) },
        { properties: { content: { type: 'array', items: { $ref: getSchemaPath(AvisResponseDto) } } } },
      ],
    },
  })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Utilisateur non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getAvisByUtilisateur(
    @Param('utilisateurId', new ParseUUIDPipe()) utilisateurId: string,
    @Query('page', new ParseIntPipe({ optional: true })) page = 0,
    @Query('size', new ParseIntPipe({ optional: true })) size = 10,
  ): Promise<PaginatedResponseDto<AvisResponseDto>> {
    return this.service.getAvisByUtilisateur(utilisateurId, page, size);
  }

  @Get('vehicule/:vehiculeId')
  @ApiOperation({ summary: 'Avis sur un véhicule' })
  @ApiParam({ name: 'vehiculeId', type: String, format: 'uuid' })
  @ApiQuery({ name: 'page', required: false, type: Number, example: 0 })
  @ApiQuery({ name: 'size', required: false, type: Number, example: 10 })
  @ApiResponse({
    status: 200,
    description: 'Avis récupérés avec succès',
    schema: {
      allOf: [
        { $ref: getSchemaPath(PaginatedResponseDto) },
        { properties: { content: { type: 'array', items: { $ref: getSchemaPath(AvisResponseDto) } } } },
      ],
    },
  })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Véhicule non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getAvisByVehicule(
    @Param('vehiculeId', new ParseUUIDPipe()) vehiculeId: string,
    @Query('page', new ParseIntPipe({ optional: true })) page = 0,
    @Query('size', new ParseIntPipe({ optional: true })) size = 10,
  ): Promise<PaginatedResponseDto<AvisResponseDto>> {
    return this.service.getAvisByVehicule(vehiculeId, page, size);
  }

  @Get('garage/:garageId')
  @ApiOperation({ summary: 'Avis sur un garage' })
  @ApiParam({ name: 'garageId', type: String, format: 'uuid' })
  @ApiQuery({ name: 'page', required: false, type: Number, example: 0 })
  @ApiQuery({ name: 'size', required: false, type: Number, example: 10 })
  @ApiResponse({
    status: 200,
    description: 'Avis récupérés avec succès',
    schema: {
      allOf: [
        { $ref: getSchemaPath(PaginatedResponseDto) },
        { properties: { content: { type: 'array', items: { $ref: getSchemaPath(AvisResponseDto) } } } },
      ],
    },
  })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Garage non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getAvisByGarage(
    @Param('garageId', new ParseUUIDPipe()) garageId: string,
    @Query('page', new ParseIntPipe({ optional: true })) page = 0,
    @Query('size', new ParseIntPipe({ optional: true })) size = 10,
  ): Promise<PaginatedResponseDto<AvisResponseDto>> {
    return this.service.getAvisByGarage(garageId, page, size);
  }

  @Get('utilisateur/:utilisateurId/moyenne')
  @ApiOperation({ summary: 'Note moyenne utilisateur' })
  @ApiParam({ name: 'utilisateurId', type: String, format: 'uuid' })
  @ApiResponse({ status: 200, description: 'Note moyenne récupérée avec succès', schema: { type: 'number', example: 4.6 } })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Utilisateur non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getNoteMoyenneUtilisateur(
    @Param('utilisateurId', new ParseUUIDPipe()) utilisateurId: string,
  ): Promise<number> {
    return this.service.getNoteMoyenneUtilisateur(utilisateurId);
  }

  @Get('vehicule/:vehiculeId/moyenne')
  @ApiOperation({ summary: 'Note moyenne véhicule' })
  @ApiParam({ name: 'vehiculeId', type: String, format: 'uuid' })
  @ApiResponse({ status: 200, description: 'Note moyenne récupérée avec succès', schema: { type: 'number', example: 4.3 } })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Véhicule non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getNoteMoyenneVehicule(@Param('vehiculeId', new ParseUUIDPipe()) vehiculeId: string): Promise<number> {
    return this.service.getNoteMoyenneVehicule(vehiculeId);
  }

  @Get('garage/:garageId/moyenne')
  @ApiOperation({ summary: 'Note moyenne garage' })
  @ApiParam({ name: 'garageId', type: String, format: 'uuid' })
  @ApiResponse({ status: 200, description: 'Note moyenne récupérée avec succès', schema: { type: 'number', example: 4.1 } })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Garage non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  getNoteMoyenneGarage(@Param('garageId', new ParseUUIDPipe()) garageId: string): Promise<number> {
    return this.service.getNoteMoyenneGarage(garageId);
  }

  @Post(':avisId/signaler')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Signaler un avis' })
  @ApiParam({ name: 'avisId', type: String, format: 'uuid' })
  @ApiResponse({
    status: 200,
    description: 'Avis signalé avec succès',
    schema: { type: 'object', properties: { message: { type: 'string', example: 'Avis signalé avec succès' } } },
  })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Données invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Avis non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  async signalerAvis(
    @Param('avisId', new ParseUUIDPipe()) avisId: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<{ message: string }> {
    await this.service.signalerAvis(avisId, user);
    return { message: 'Avis signalé avec succès' };
  }

  @Delete(':avisId')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Supprimer un avis' })
  @ApiParam({ name: 'avisId', type: String, format: 'uuid' })
  @ApiResponse({ status: 204, description: 'Avis supprimé avec succès' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 403, type: ApiErrorResponseDto, description: 'Interdit - Accès refusé' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Avis non trouvé' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  async deleteAvis(
    @Param('avisId', new ParseUUIDPipe()) avisId: string,
    @CurrentUser() user: AuthenticatedUser,
  ): Promise<void> {
    await this.service.deleteAvis(avisId, user);
  }

  @Get('transaction/:transactionId/validation')
  @ApiOperation({ summary: 'Valider transaction pour avis' })
  @ApiParam({ name: 'transactionId', type: String, format: 'uuid' })
  @ApiQuery({ name: 'typeAvis', type: String, required: true, example: 'ACHETEUR' })
  @ApiResponse({ status: 200, description: 'Validation récupérée avec succès', schema: { type: 'boolean', example: true } })
  @ApiResponse({ status: 400, type: ApiErrorResponseDto, description: 'Paramètres invalides' })
  @ApiResponse({ status: 401, type: ApiErrorResponseDto, description: 'Non autorisé - Token invalide ou expiré' })
  @ApiResponse({ status: 404, type: ApiErrorResponseDto, description: 'Transaction non trouvée' })
  @ApiResponse({ status: 500, type: ApiErrorResponseDto, description: 'Erreur serveur interne' })
  isTransactionValide(
    @Param('transactionId', new ParseUUIDPipe()) transactionId: string,
    @Query('typeAvis') typeAvis: string,
  ): Promise<boolean> {
    return this.service.isTransactionValide(transactionId, typeAvis);
  }
}
